Foundation/NSURLProtocol.swift

```diff
@@ -323,19 +435,62 @@ open class URLProtocol : NSObject {
         The only way that failure can occur is if the given class is not a
         subclass of NSURLProtocol.
     */
-    open class func registerClass(_ protocolClass: AnyClass) -> Bool { NSUnimplemented() }
-    
+    open class func registerClass(_ protocolClass: AnyClass) -> Bool {
+        if protocolClass is URLProtocol.Type {
+            _classesLock.lock()
+            guard !_registeredProtocolClasses.contains(where: { $0 === protocolClass }) else {
+                _classesLock.unlock()
+                return true
+            }
+            _registeredProtocolClasses.append(protocolClass)
+            _classesLock.unlock()
+            return true
+        }
+        return false
+    }
+
+    // TODO: Synchronize all the class funcs that use the registered protocol classes array
+    internal class func getProtocolClass(for urlRequest: URLRequest) -> AnyClass? {
+        // Registered protocols are consulted in reverse order.
+        // This behaviour makes custom protocols to be consulted first, if any.
+        _classesLock.lock()
+        let protocolClasses  = _registeredProtocolClasses.reversed()
+        for protocolClass in protocolClasses {
+            let urlProtocolClass : AnyClass = protocolClass
+            if (urlProtocolClass as! URLProtocol.Type).canInit(with: urlRequest) {
+	        _classesLock.unlock()
+                return urlProtocolClass
+            }
+        }
+        _classesLock.unlock()
+        return nil
+    }
+
     /*! 
         @method unregisterClass:
         @abstract This method unregisters a protocol. 
         @discussion After unregistration, a protocol class is no longer
         consulted in calls to NSURLProtocol class methods.
         @param protocolClass The class to unregister.
     */
-    open class func unregisterClass(_ protocolClass: AnyClass) { NSUnimplemented() }
+    open class func unregisterClass(_ protocolClass: AnyClass) {
+        if let idx = _registeredProtocolClasses.index(where: { $0 === protocolClass }) {
+            _classesLock.lock()
```

Comment:
The lock acquisition needs to be raised above the `if`, given that `_registeredProtocolClasses` may be concurrently accessed